vm/systemSmartContracts/staking.go

```diff
@@ -0,0 +1,226 @@
+package systemSmartContracts
+
+import (
+	"bytes"
+	"encoding/json"
+	"math/big"
+
+	"github.com/ElrondNetwork/elrond-go/core/logger"
+	"github.com/ElrondNetwork/elrond-go/vm"
+	vmcommon "github.com/ElrondNetwork/elrond-vm-common"
+)
+
+var log = logger.DefaultLogger()
+
+const ownerKey = "owner"
+
+type stakingData struct {
+	StartNonce    uint64   `json:"StartNonce"`
+	Staked        bool     `json:"Staked"`
+	UnStakedNonce uint64   `json:"UnStakedNonce"`
+	BlsPubKey     []byte   `json:"BlsPubKey"`
+	StakeValue    *big.Int `json:"StakeValue"`
+}
+
+type stakingSC struct {
+	eei        vm.SystemEI
+	stakeValue *big.Int
+}
+
+// NewStakingSmartContract creates a staking smart contract
+func NewStakingSmartContract(stakeValue *big.Int, eei vm.SystemEI) (*stakingSC, error) {
+	if stakeValue == nil {
+		return nil, vm.ErrNilInitialStakeValue
+	}
+	if eei == nil || eei.IsInterfaceNil() {
+		return nil, vm.ErrNilSystemEnvironmentInterface
+	}
+
+	reg := &stakingSC{stakeValue: big.NewInt(0).Set(stakeValue), eei: eei}
+	return reg, nil
+}
+
+// Execute calls one of the functions from the staking smart contract and runs the code according to the input
+func (r *stakingSC) Execute(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	if CheckIfNil(args) != nil {
+		return vmcommon.UserError
+	}
+
+	switch args.Function {
+	case "_init":
+		return r.init(args)
+	case "stake":
+		return r.stake(args)
+	case "unStake":
+		return r.unStake(args)
+	case "finalizeUnStake":
+		return r.finalizeUnStake(args)
+	case "slash":
+		return r.slash(args)
+	}
+
+	return vmcommon.UserError
+}
+
+func (r *stakingSC) init(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	r.eei.SetStorage([]byte(ownerKey), args.CallerAddr)
+	r.eei.SetStorage(args.CallerAddr, big.NewInt(0).Bytes())
+	return vmcommon.Ok
+}
+
+func (r *stakingSC) stake(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	if args.CallValue.Cmp(r.stakeValue) != 0 {
+		return vmcommon.UserError
+	}
+
+	registrationData := stakingData{
+		StartNonce:    0,
+		Staked:        false,
+		BlsPubKey:     nil,
+		UnStakedNonce: 0,
+		StakeValue:    big.NewInt(0),
+	}
+	data := r.eei.GetStorage(args.CallerAddr)
+
+	if data != nil {
+		err := json.Unmarshal(data, registrationData)
```

Comment:
might want to inject marshalizer interface instead of using json

Reply:
It will be much harder to move the code to another VM.